Publish @open-spaced-repetition/srs-kit as a public package (BSD-3-Clause) (#418)
open-spaced-repetition/ts-fsrs: packages/srs-kit/README.md

# @open-spaced-repetition/srs-kit

> Shared SRS framework primitives, schema helpers, model contracts, middleware helpers, and scheduler building blocks for the Open Spaced Repetition `ts-fsrs` project.

Reviewer: > Shared SRS framework primitives, schema helpers, model contracts

I'm not sure what this really means. If this is just me being dumb ignore the suggestion.

```suggestion
Shared SRS functionality for the Open Spaced Repetition `ts-fsrs` project.
```

Author: I probably should clarify the scope of `srs-kit`.

I define `srs-kit` as the architectural foundation layer that supports `ts-fsrs v6` (https://github.com/open-spaced-repetition/ts-fsrs/issues/373#issuecomment-4777701801).

The purpose of `srs-kit` is not to provide a standalone SRS algorithm implementation, but to provide the type-safe infrastructure layer required to build `ts-fsrs` v6.

Currently, `srs-kit` provides:

- Type-safe framework primitives
- Model contracts and model declaration APIs
- Schema helpers
- Date/time handling abstractions (`chrono`)
- Scheduler declaration APIs with full type inference
- Middleware composition and scheduler building blocks

The goal is to make it possible to define the scheduler architecture like this:

```ts
const Scheduler = defineScheduler({
  model: FSRS6Model,
  chrono: dateChrono,
}).use(
  FuzzingMiddleware,
  StatsMiddleware,
  ScheduledDaysMiddleware,
  LearningStepsMiddleware,
  MonotonicIntervalMiddleware,
)

const schedulerInstance = Scheduler.create({ config })

const record = schedulerInstance.review({
  card,
  grade: Rating.Good,
  now: new Date(),
})
```

For example:

- When using `numericChrono`, `review.now` can simply accept elapsed days instead of a `Date`.

- When `LeechMiddleware` is injected, the scheduler can automatically add `scheduleStatus = "suspend"` after the card reaches the configured lapse threshold.  https://github.com/open-spaced-repetition/ts-fsrs/blob/c3b5f8718a13d135ee46795cb7d4547a0b98f59e/packages/fsrs/src/middlewares/leech/middleware.spec.ts
<img width="1536" height="728" alt="image" src="https://github.com/user-attachments/assets/a428e69c-d857-419f-856e-800c4978a3b8" />

- When `LearningStepsMiddleware` is injected, the middleware can extend the required schema so `config` requires `learningSteps[]` / `relearningSteps[]`, and cards require fields such as `step`.  https://github.com/open-spaced-repetition/ts-fsrs/blob/ef928947369d9ec616be256561288b0491c9cea7/packages/fsrs/src/middlewares/learning-steps/middleware.spec.ts
<img width="1868" height="547" alt="image" src="https://github.com/user-attachments/assets/ed9c2a5d-e097-45dc-974b-7cbee1be1424" />

The main purpose of `srs-kit` is to provide a highly extensible and type-safe foundation through schema composition and TypeScript type inference, allowing `ts-fsrs` to build a more modular scheduler architecture.

Reviewer: Maybe you can just add your explanation comment to the readme? I'm not sure how important the readme is if its just a package that is used internally by ts-fsrs 😆

Author: > Maybe you can just add your explanation comment to the readme? I'm not sure how important the readme is if its just a package that is used internally by ts-fsrs 😆

My concern is that people might see `srs-kit` and start using it directly as a standalone framework for their own projects 🤣

Reviewer: ```suggestion
<blockquote>
  <strong>⚠️ WARNING:</strong> This package is only intended for use as a dependency of the `ts-fsrs` package. Please use that package instead.
</blockquote>

Shared SRS framework primitives, schema helpers, model contracts, middleware helpers, and scheduler building blocks for the Open Spaced Repetition `ts-fsrs` project.
```
> [!CAUTION]
> We could use the GitHub style warnings but I doubt that will work on the NPM homepage / other places you might want to copy this.

## Development status

`srs-kit` is currently under active development. It is published as a public package so `ts-fsrs` and related packages can consume the shared implementation while the framework is being refined.

> The APIs in this package should be treated as unstable. Method names, function signatures, exports, types, schemas, scheduler contracts, and runtime behavior may change in future releases, including changes announced through changesets. Do not assume compatibility across versions unless a future release explicitly documents a stability guarantee.

Reviewer: ```suggestion
```

I guess if we include this it implies people should use the package?

Author: > I guess if we include this it implies people should use the package?

🤔 That's a good point.
